micronaut-maven-plugin/src/main/java/io/micronaut/maven/ImportFactoryMojo.java

```diff
@@ -0,0 +1,264 @@
+
+    if (dependencies.isEmpty()) {
+      getLog().warn("No matching dependencies.");
+      return;
+    }
+
+    getLog().info("Found " + dependencies.size() + " matching dependencies:");
+    dependencies.forEach(
+        dependency -> getLog().info(getIdentifier(dependency) + " at " + dependency.getFile()));
+
+    List<String> packages = getFilteredPackages(dependencies);
+
+    if (packages.isEmpty()) {
+      getLog().warn("No matching packages.");
+      return;
+    }
+
+    getLog().info("Found " + packages.size() + " matching packages:");
+    packages.forEach(getLog()::info);
+
+    if (targetPackage == null || targetPackage.isEmpty()) {
+      for (String packageName : packages) {
+        try {
+          generateImportFactory(packageName, Collections.singletonList(packageName));
+        } catch (IOException e) {
+          throw new MojoExecutionException("Error creating factory for " + packageName, e);
+        }
+      }
+    } else {
+      try {
+        generateImportFactory(targetPackage, packages);
+      } catch (IOException e) {
+        throw new MojoExecutionException("Error creating factory for " + targetPackage, e);
+      }
+    }
+  }
+
+  /**
+   * Get the list of matching packages from the given list of dependencies.
+   *
+   * @param dependencies the Maven dependencies
+   * @return a list of matching packages
+   * @throws MojoExecutionException when the artifacts can't be read
+   */
+  private List<String> getFilteredPackages(List<Artifact> dependencies)
+      throws MojoExecutionException {
+    List<String> packages = new ArrayList<>();
+    for (Artifact dependency : dependencies) {
+      packages.addAll(getPackages(dependency));
+    }
+
+    Pattern includePackages = Pattern.compile(includePackagesFilter);
+    Pattern excludePackages = Pattern.compile(excludePackagesFilter);
+
+    return packages.stream().filter(includePackages.asMatchPredicate())
+        .filter(excludePackages.asMatchPredicate().negate()).sorted().toList();
+  }
+
+  /**
+   * Get the list of matching dependencies.
+   *
+   * @return the matching dependencies
+   */
+  private List<Artifact> getFilteredDependencies() {
+    Pattern includeDependency = Pattern.compile(includeDependenciesFilter);
+    Pattern excludeDependency = Pattern.compile(excludeDependenciesFilter);
+
+    return project.getArtifacts().stream()
+        .filter(dependency -> includeDependency.matcher(getIdentifier(dependency)).matches())
+        .filter(dependency -> !excludeDependency.matcher(getIdentifier(dependency)).matches())
+        .toList();
+  }
+
+  /**
+   * Get the packages from the given artifact.
+   *
+   * @param artifact the Maven artifact
+   * @return the list of packages
+   * @throws MojoExecutionException when the artifacts can't be read
+   */
+  private List<String> getPackages(Artifact artifact) throws MojoExecutionException {
+    try (JarFile file = new JarFile(artifact.getFile(), false)) {
```

Comment:
I wonder about the performance of this method. When using this feature, we're essentially reading every JAR every time.

Perhaps we should cache these results by hashing on the GAV or the artifact path.

Reply:
I recon it will be very fast since we are only iterating over the files in the jar (entries in the zipfile) and not actually unpacking them. You nevertheless probably don't want this to be invoked on all dependencies, hence the `includeDependenciesFilter` and `excludeDependenciesFilter` parameters.

Is there any infrastructure for caching, or do we have to implement that ourselves?